### PR TITLE
Add aspnet/Extensions build.

### PR DESCRIPTION
**Pull request by crummel**

### CODE REVIEW

**Comment by crummel** [comment]
cc @natemcmaster

File: eng/Version.Details.xml [hunk 1]
```diff
@@ -49,6 +49,10 @@
       <Sha>27af96bdb7ba8d6d7ea9ad53fc76cd1d1aa80703</Sha>
       <RepoName>nuget-client</RepoName>
     </Dependency>
+    <Dependency Name="Microsoft.Extensions.Logging" Version="3.0.0-rtm.19279.9">
```
Comment (natemcmaster) [inline eng/Version.Details.xml]: Where did this version number come from? The latest build should be "3.0.0-preview6.19279.7"

File: repos/aspnet-extensions.proj [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  <PropertyGroup>
+    <OfficialBuildId>20190529.9</OfficialBuildId>
+
+    <BuildCommandArgs/>
+    <BuildCommandArgs>$(BuildCommandArgs) --configuration $(Configuration)</BuildCommandArgs>
+    <BuildCommandArgs>$(BuildCommandArgs) --binaryLog</BuildCommandArgs>
+    <BuildCommand>$(ProjectDirectory)build$(ShellExtension) $(BuildCommandArgs)</BuildCommand>
+
+    <PackagesOutput>$(ProjectDirectory)artifacts/packages/$(Configuration)/Shipping/</PackagesOutput>
```
Comment (natemcmaster) [inline repos/aspnet-extensions.proj]: There are going to be some packages in `artifacts/packages/$(Configuration)/NonShipping/` as well which are used to transport bits upstream to other repos. Is it possible to define multiple package output paths?

File: repos/core-setup.proj [hunk 3]
```diff
@@ -34,6 +34,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <RepositoryReference Include="aspnet-extensions" />
```
Comment (natemcmaster) [inline repos/core-setup.proj]: aspnet-extensions comes after core-setup. The dependency flow of repos is:

```
dotnet/core-setup
 -> aspnet/Extensions
    -> aspnet/AspNetCore-Tooling 
        -> dotnet/toolset
        -> aspnet/AspNetCore
    -> aspnet/AspNetCore 
        -> aspnet/websdk
        -> dotnet/core-sdk
```

File: eng/Version.Details.xml (hunk 1 above)
Comment (crummel) [reply eng/Version.Details.xml]: Fixed this.

File: repos/aspnet-extensions.proj (hunk 2 above)
Comment (crummel) [reply repos/aspnet-extensions.proj]: Not currently, but I think we should be able to workaround this using the output placement API - I'm giving this a try now.

File: repos/core-setup.proj (hunk 3 above)
Comment (crummel) [reply repos/core-setup.proj]: Fixed the repo dependencies.

File: eng/Version.Details.xml [hunk 4]
```diff
@@ -105,7 +109,7 @@
       <Uri>https://github.com/dotnet/cli-migrate</Uri>
       <Sha>37e3a01aa1df320a286b0a858876f9994fb6bb18</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.Extensions.CommandLineUtils" Version="1.0.1">
+    <Dependency Name="Microsoft.Extensions.CommandLineUtils" Version="3.0.0-preview5.19227.9">
```
Comment (natemcmaster) [inline eng/Version.Details.xml]: There is no such package. This was deprecated in 1.1. There won't be a new 3.0 version of this.

**Comment by crummel** [comment]
Hi @JunTaoLuo, after the SDK and Arcade update I've been making some good progress on this.  Right now what I'm running into is
```
  /home/chris/dotnet_source-build/bin/src/Extensions.48cb2abb441bfa3affbcd9ae0988af05b2416e03/src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj : error NU1202: Package Microsoft.CodeAnalysis.Common 3.1.0-dev-19067-14 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3). Package Microsoft.CodeAnalysis.Common 3.1.0-dev-19067-14 supports: netstandard2.0 (.NETStandard,Version=v2.0) [/home/chris/dot
net_source-build/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /home/chris/dotnet_source-build/bin/src/Extensions.48cb2abb441bfa3affbcd9ae0988af05b2416e03/src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj : error NU1202: Package Microsoft.CodeAnalysis.CSharp 3.1.0-dev-19067-14 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3). Package Microsoft.CodeAnalysis.CSharp 3.1.0-dev-19067-14 supports: netstandard2.0 (.NETStandard,Version=v2.0) [/home/chris/dot
net_source-build/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /home/chris/dotnet_source-build/bin/src/Extensions.48cb2abb441bfa3affbcd9ae0988af05b2416e03/src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj : error NU1202: Package Microsoft.CodeAnalysis.CSharp.Workspaces 3.1.0-dev-19067-14 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3). Package Microsoft.CodeAnalysis.CSharp.Workspaces 3.1.0-dev-19067-14 supports: netstandard2.0 (.NETStandard,Version=
v2.0) [/home/chris/dotnet_source-build/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /home/chris/dotnet_source-build/bin/src/Extensions.48cb2abb441bfa3affbcd9ae0988af05b2416e03/src/Logging/Logging.Analyzers/src/Microsoft.Extensions.Logging.Analyzers.csproj : error NU1202: Package Microsoft.CodeAnalysis.Common 3.1.0-dev-19067-14 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3). Package Microsoft.CodeAnalysis.Common 3.1.0-dev-19067-14 supports: netstandard2.0 (.NETStandard,Version=v2.0)
[/home/chris/dotnet_source-build/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /home/chris/dotnet_source-build/bin/src/Extensions.48cb2abb441bfa3affbcd9ae0988af05b2416e03/src/Logging/Logging.Analyzers/src/Microsoft.Extensions.Logging.Analyzers.csproj : error NU1202: Package Microsoft.CodeAnalysis.CSharp 3.1.0-dev-19067-14 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3). Package Microsoft.CodeAnalysis.CSharp 3.1.0-dev-19067-14 supports: netstandard2.0 (.NETStandard,Version=v2.0)
[/home/chris/dotnet_source-build/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /home/chris/dotnet_source-build/bin/src/Extensions.48cb2abb441bfa3affbcd9ae0988af05b2416e03/src/Logging/Logging.Analyzers/src/Microsoft.Extensions.Logging.Analyzers.csproj : error NU1202: Package Microsoft.CodeAnalysis.CSharp.Workspaces 3.1.0-dev-19067-14 is not compatible with netstandard1.3 (.NETStandard,Version=v1.3). Package Microsoft.CodeAnalysis.CSharp.Workspaces 3.1.0-dev-19067-14 supports: netstandard2.0 (.NETS
tandard,Version=v2.0) [/home/chris/dotnet_source-build/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
```

i.e., Microsoft.DotNet.Analyzers.Async is built for netstandard1.3 but in the newer Roslyn build, the packages it relies on is only built for netstandard2.0.  I'm trying to patch in netstandard1.3 support to Roslyn now but if that doesn't work is it possible to upgrade?

**Comment by crummel** [comment]
Another option we've done before is having the target framework be an overridable property that we set in source-build, that's probably the more flexible option.

**Comment by JunTaoLuo** [comment]
I'll try to help with this investigation today but the error doesn't ring a bell at the moment.

**Comment by JunTaoLuo** [comment]
I spoke with @anurse  and let's not go down the rabbit hole of adding netstandard1.3 support @crummel . Can you exclude building Microsoft.Extensions.Logging.Analyzers in source build? It's not a package we include in the framework so we don't need strictly need to build it. That being said, there's no reason why we shouldn't be able to build it and we could look into updating it to netstandard2.0 but let's not block on this decision.

TL;DR, skip building it for now and we'll investigate upgrading it to netstandar2.0

**Comment by crummel** [comment]
Perfect, we can just add `<ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>` to that project, I'll patch it in for now.  I think I'm down to one last error:
```
  /home/chris/dotnet_source-build/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Collections.Immutable" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [/home/chris/dotnet_source-build/bin/src/Extensions.48cb2abb441bfa3affbcd9ae0988af05b2416e03/src/TargetingPack/Microsoft.Internal.Extensions.Refs/ref/Microsoft.Internal.Extensions.Refs.csproj]
  /home/chris/dotnet_source-build/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Collections.Immutable" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [/home/chris/dotnet_source-build/bin/src/Extensions.48cb2abb441bfa3affbcd9ae0988af05b2416e03/src/TargetingPack/Microsoft.Internal.Extensions.Refs/ref/Microsoft.Internal.Extensions.Refs.csproj]
```

Does `Internal` mean we can exclude this one too?  😃

**Comment by JunTaoLuo** [comment]
https://github.com/aspnet/Extensions/blob/master/src/TargetingPack/Microsoft.Internal.Extensions.Refs/ref/Microsoft.Internal.Extensions.Refs.csproj#L6

We don't ship it, so skip that too.

**Comment by JunTaoLuo** [comment]
Hmm actually, I just looked at this package again and this is used to build our reference packages for Extensions and AspNetCore. To my understanding, we would need to build the reference assemblies right? If so, we'll need to resolve the error. Any more details on why there are conflicts with System.Collections.Immutable?

**Comment by crummel** [comment]
Ok, I put that project back in and I'm looking into the System.Immutable.Collections issue.  It looks like the hash used in preview6 was actually [54d51a34](https://github.com/aspnet/Extensions/commit/54d51a340698b6883dd3e47be372c07e0acf75bc), does that seem right?  Usually we run into this issue when a repo is incoherent with the others, I can try that hash and see if it fixes the issue.

**Comment by crummel** [comment]
Beyond the TargetFramework overrides, once we know we have the correct set of packages we'll want to add `ExcludeFromSourceBuild` tags in your repo.  The other patches are all for source-build issues so that should be it on the ASP.Net side.

**Comment by JunTaoLuo** [comment]
Yes that's the commit hash of our preview 6 release.

**Comment by crummel** [comment]
This just needs prebuilt fixes now, ready for review.

**Comment by crummel** [comment]
This is unfortunately quite an increase in prebuilts - the System.* references might be able to be consolidated with existing versions.  I'm trying to work around adding the new Roslyn packages, since that's adding a SQLite reference we definitely don't want.

@JunTaoLuo - do you know for these if they're just used as reference packages?
- System.Buffers.4.5.0
- System.ComponentModel.Annotations.4.6.0-preview6.19303.8
- System.Diagnostics.DiagnosticSource.4.6.0-preview6.19303.8
- System.Diagnostics.EventLog.4.6.0-preview6.19303.8
- System.Memory.4.5.3
- System.Numerics.Vectors.4.5.0
- System.Runtime.CompilerServices.Unsafe.4.6.0-preview6.19303.8
- System.Security.AccessControl.4.6.0-preview6.19303.8
- System.Security.Cryptography.Cng.4.6.0-preview6.19303.8
- System.Security.Cryptography.Pkcs.4.6.0-preview6.19303.8
- System.Security.Cryptography.Xml.4.6.0-preview6.19303.8
- System.Security.Permissions.4.6.0-preview6.19303.8
- System.Security.Principal.Windows.4.6.0-preview6.19303.8
- System.Text.Encodings.Web.4.6.0-preview6.19303.8
- System.Text.Json.4.6.0-preview6.19303.8

**Comment by crummel** [comment]
@JunTaoLuo also, do you think we could get rid of these projects?  I think that would get rid of the new Roslyn dependencies.
- src/Analyzers/Async/src/Microsoft.DotNet.Analyzers.Async.csproj
- src/Logging/Logging.Analyzers/src/Microsoft.Extensions.Logging.Analyzers.csproj

**Comment by JunTaoLuo** [comment]
Yes we can definitely get rid of the analyzers, they are not in the shared framework. Let me check the System.* packages.

**Comment by JunTaoLuo** [comment]
Is this ready to merge now that all checks are green?

**Comment by crummel** [comment]
I think this is ready but I was hoping for reviews from @adaggarwal and/or @dseefeld.

**Comment by crummel** [comment]
Merging this ahead of @dseefeld's review since we have some conflicting OOF times and this blocks AspNetCore-Tooling.  Let me know if there's any changes I should make.

Remaining aspnet/Extensions work:
- Long-term TargetFramework fix.
- Incorporate `ExcludeFromSourceBuild` patches.
- Prebuilt reduction.

Remaining source-build work:
- Uptake new Extensions after patch incorporation.
